### PR TITLE
Fix: path item objects become block scalars [#3557]

### LINKED ISSUE #3557: Generated swagger.yaml is corrupted in openapiv2 v2.17.0+: Path item becomes string instead of object
## 🐛 Bug Report

Starting from version v2.17.0 of openapiv2, the generated swagger.yaml file is corrupted. Specifically, where the path item should be an object, it has become a string.

## To Reproduce

Execute the following using openapiv2 v2.17.1:

```sh
protoc -I . --openapiv2_out ./gen \
    --openapiv2_opt=output_format=yaml \
    your/service/v1/your_service.proto
```

For this, let's assume your_service.proto is as follows:

```proto
syntax = "proto3";
package your.service.v1;

import "google/api/annotations.proto";

option go_package = "github.com/yourorg/yourprotos/gen/go/your/service/v1";

message StringMessage {
  string value = 1;
}

service YourService {
  rpc Echo(StringMessage) returns (StringMessage) {
    option (google.api.http) = {post: "/api/echo"};
  }
}
```

## Expected behavior

The file generated with openapiv2 v.2.6.2 is as follows:

```yaml
swagger: "2.0"
info:
  title: your/service/v1/your_service.proto
  version: version not set
tags:
  - name: YourService
consumes:
  - application/json
produces:
  - application/json
paths:
  /api/echo:
    post:
      operationId: YourService_Echo
      responses:
        "200":
          description: A successful response.
          schema:
            $ref: '#/definitions/v1StringMessage'
        default:
          description: An unexpected error response.
          schema:
            $ref: '#/definitions/rpcStatus'
      parameters:
        - name: value
          in: query
          required: false
          type: string
      tags:
        - YourService
definitions:
  protobufAny:
    type: object
    properties:
      '@type':
        type: string
    additionalProperties: {}
  rpcStatus:
    type: object
    properties:
      code:
        type: integer
        format: int32
      message:
        type: string
      details:
        type: array
        items:
          type: object
          $ref: '#/definitions/protobufAny'
  v1StringMessage:
    type: object
    properties:
      value:
        type: string
```

## Actual Behavior

The file generated with openapiv2 v2.17.1 is as follows. The path item corresponding to /api/echo has turned into a string.

```yaml
swagger: "2.0"
info:
  title: your/service/v1/your_service.proto
  version: version not set
tags:
  - name: YourService
consumes:
  - application/json
produces:
  - application/json
paths:
  /api/echo: |
    post:
        operationId: YourService_Echo
        responses:
            "200":
                description: A successful response.
                schema:
                    $ref: '#/definitions/v1StringMessage'
            default:
                description: An unexpected error response.
                schema:
                    $ref: '#/definitions/rpcStatus'
        parameters:
            - name: value
              in: query
              required: false
              type: string
        tags:
            - YourService
definitions:
  protobufAny:
    type: object
    properties:
      '@type':
        type: string
    additionalProperties: {}
  rpcStatus:
    type: object
    properties:
      code:
        type: integer
        format: int32
      message:
        type: string
      details:
        type: array
        items:
          type: object
          $ref: '#/definitions/protobufAny'
  v1StringMessage:
    type: object
    properties:
      value:
        type: string
```

## Your Environment

Mac OS - Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_X86_64 x86_64

### PR DESCRIPTION
**Pull request by qawatake**
#### References to other Issues or PRs

Fixes #3557

#### Have you read the [Contributing Guidelines](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/CONTRIBUTING.md)?

Yes.

#### Brief description of what is fixed or changed

Fixes a bug where the path item object is marshaled as a block scalar in openapiv2 v2.17.0 and later.

```yaml
swagger: "2.0"
info:
  title: your/service/v1/your_service.proto
  version: version not set
tags:
  - name: YourService
consumes:
  - application/json
produces:
  - application/json
paths:
  /api/echo: |
    post:
        operationId: YourService_Echo
        responses:
            "200":
                description: A successful response.
                schema:
                    $ref: '#/definitions/v1StringMessage'
            default:
                description: An unexpected error response.
                schema:
                    $ref: '#/definitions/rpcStatus'
        parameters:
            - name: value
              in: query
              required: false
              type: string
        tags:
            - YourService
definitions:
~~ snip ~~
```

### Summary by CodeRabbit

- Bug Fix: Updated the `MarshalYAML` method in `generator.go` to use a new `toYAMLNode` method for converting `PathItemObject` to a YAML node. This change addresses a bug in the go-yaml library.
- Test: Added a new test function `TestGenerateYAML` in `generator_test.go` to validate the generation of YAML files from protobuf input.
- New Feature: Introduced a new protobuf file `path_item_object.prototext` for testing purposes, which includes a service definition and a message type.
- Chore: Modified the `go_test` rule in `BUILD.bazel` to include all files under the `testdata` directory during the build process.

### CODE REVIEW

File: protoc-gen-openapiv2/internal/genopenapi/generator.go [hunk 1]
```diff
@@ -150,20 +150,45 @@ func (po openapiPathsObject) MarshalYAML() (interface{}, error) {
...
+//	}
+func (pio *openapiPathItemObject) toYAMLNode() (*yaml.Node, error) {
+	var doc yaml.Node
+	var buf bytes.Buffer
+	ec := yaml.NewEncoder(&buf)
+	ec.SetIndent(2)
+	if err := ec.Encode(pio); err != nil {
+		return nil, err
+	}
+	if err := yaml.Unmarshal(buf.Bytes(), &doc); err != nil {
```
Comment (qawatake) [inline protoc-gen-openapiv2/internal/genopenapi/generator.go:183]: I applied the same workaround as in https://github.com/grpc-ecosystem/grpc-gateway/pull/2801.

File: protoc-gen-openapiv2/internal/genopenapi/generator.go [hunk 2]
```diff
@@ -150,20 +150,45 @@ func (po openapiPathsObject) MarshalYAML() (interface{}, error) {
...
+		return nil, err
+	}
+	if len(doc.Content) == 0 {
+		return nil, errors.New("unexpected number of yaml nodes")
+	}
+	return doc.Content[0], nil
```
[the remark this answers is not in the document]
Comment (qawatake) [reply protoc-gen-openapiv2/internal/genopenapi/generator.go:189]: `(n *Node) Encode` in go-yaml is implemented in the same manner.

https://github.com/go-yaml/yaml/blob/f6f7691b1fdeb513f56608cd2c32c51f8194bf51/yaml.go#L259

[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply protoc-gen-openapiv2/internal/genopenapi/generator.go:189]: Yeah this comment is wrong, this is fine :)

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Hi, thanks for this PR! Any chance you could add a test file that exhibits the old, bad behaviour, so we know this is now fixed? Thanks!

**Comment by qawatake** [comment]
I've added a test. However, the `textproto` file used for the test data might be a bit too large since it was auto-generated. What do you think?

**Review by johanbrandhorst — APPROVED** [review]

**Comment by johanbrandhorst** [comment]
It looks like you need to regenerate one of the bazel files, there are instructions in CONTRIBUTING.md, thanks!

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Regenerate bazel files

**Review by johanbrandhorst — APPROVED** [review]
LGTM, thank you!

**Pull request merged**